File: bindings/imgui_bundle/demos_python/sandbox/sandbox_tstengine.py
from imgui_bundle import imgui, hello_imgui
from imgui_bundle.imgui.test_engine_checks import CHECK

test_open_metric: imgui.test_engine.Test
test_capture_screenshot: imgui.test_engine.Test


def my_register_tests():
    global test_open_metric, test_capture_screenshot
    engine = hello_imgui.get_imgui_test_engine()

    # test with gui and vars
    test_with_vars = imgui.test_engine.register_test(engine, "demo_tests", "test_vars")
    class TestVar2:
        my_int = 42
    test_var2 = TestVar2()
    def test_with_vars_gui_func(ctx: imgui.test_engine.TestContext):
        imgui.set_next_window_size(hello_imgui.em_to_vec2(40, 8))
        imgui.begin("Test Window", None, imgui.WindowFlags_.no_saved_settings)
        _, test_var2.my_int = imgui.slider_int("Slider", test_var2.my_int, 0, 1000)
        imgui.end()
    def test_with_vars_test_func(ctx: imgui.test_engine.TestContext):
        # Optional: reset test_var2 to its startup values
        nonlocal test_var2
        test_var2 = TestVar2()
        # Run the test
        ctx.set_ref("Test Window")
        CHECK(test_var2.my_int == 42)
        ctx.item_input_value("Slider", 123)
        CHECK(test_var2.my_int == 123)
    test_with_vars.test_func = test_with_vars_test_func
    test_with_vars.gui_func = test_with_vars_gui_func

    # Open Metrics
    test_open_metric = imgui.test_engine.register_test(engine, "demo_tests", "open_metrics")
    def tst_func_metrics(ctx: imgui.test_engine.TestContext):
        # This function will be executed in a separate coroutine thread which is launched by C++
        # The coroutine thread never executes in parallel with the main thread.
        # However, since it might be difficult to place a breakpoint in this thread
        ctx.yield_()
        ctx.set_ref("Dear ImGui Demo")
        ctx.menu_check("Tools/Metrics\\/Debugger")
    test_open_metric.test_func = tst_func_metrics

    # Capture entire Dear ImGui Demo window.
    test_capture_screenshot = imgui.test_engine.register_test(engine, "demo_tests", "capture_screenshot")
    def tst_func_capture(ctx: imgui.test_engine.TestContext):
        ctx.set_ref("Dear ImGui Demo")
        ctx.item_open("Widgets")
        ctx.item_open_all("Basic")
        ctx.capture_screenshot_window("Dear ImGui Demo") # not available yet!:  imgui.test_engine. ImGuiCaptureFlags_StitchAll | ImGuiCaptureFlags_HideMouseCursor);
    test_capture_screenshot.test_func = tst_func_capture


def app_gui():
    imgui.show_demo_window()
    imgui.test_engine.show_test_engine_windows(hello_imgui.get_imgui_test_engine(), None)
    if imgui.button("Run open metric automation"):
        imgui.test_engine.queue_test(hello_imgui.get_imgui_test_engine(), test_open_metric)
    if imgui.button("Run capture screenshot automation"):
        imgui.test_engine.queue_test(hello_imgui.get_imgui_test_engine(), test_capture_screenshot)


def main():
    runner_params = hello_imgui.RunnerParams()
    runner_params.use_imgui_test_engine = True

    runner_params.callbacks.show_gui = app_gui
    runner_params.callbacks.register_tests = my_register_tests

    hello_imgui.run(runner_params)


if __name__ == "__main__":
    main()

File: bindings/imgui_bundle/demos_cpp/sandbox/sandbox_tstengine.cpp
#ifdef IMGUI_BUNDLE_WITH_TEST_ENGINE
#include "imgui.h"
#include "hello_imgui/hello_imgui.h"
#include "imgui_test_engine/imgui_te_ui.h"
#include "imgui_test_engine/imgui_te_engine.h"
#include "imgui_test_engine/imgui_te_context.h"


ImGuiTest *test_open_metric, *test_capture_screenshot;


void MyRegisterTests()
{
    auto engine = HelloImGui::GetImGuiTestEngine();

    //-----------------------------------------------------------------
    // ## Open Metrics window
    //-----------------------------------------------------------------

    test_open_metric = IM_REGISTER_TEST(engine, "demo_tests", "open_metrics");
    test_open_metric->TestFunc = [](ImGuiTestContext* ctx)
    {
        ctx->SetRef("Dear ImGui Demo");
        ctx->MenuCheck("Tools/Metrics\\/Debugger");
    };

    //-----------------------------------------------------------------
    // ## Capture entire Dear ImGui Demo window.
    //-----------------------------------------------------------------

    test_capture_screenshot = IM_REGISTER_TEST(engine, "demo_tests", "capture_screenshot");
    test_capture_screenshot->TestFunc = [](ImGuiTestContext* ctx)
    {
        ctx->SetRef("Dear ImGui Demo");
        ctx->ItemOpen("Widgets");       // Open collapsing header
        ctx->ItemOpenAll("Basic");      // Open tree node and all its descendant
        ctx->CaptureScreenshotWindow("Dear ImGui Demo", ImGuiCaptureFlags_StitchAll | ImGuiCaptureFlags_HideMouseCursor);
    };

}

void AppGui()
{
    auto engine = HelloImGui::GetImGuiTestEngine();
    ImGui::ShowDemoWindow();
    ImGuiTestEngine_ShowTestEngineWindows(engine, NULL);

    if (ImGui::Button("Run open metric automation"))
        ImGuiTestEngine_QueueTest(engine, test_open_metric);
    if (ImGui::Button("Run capture screenshot automation"))
        ImGuiTestEngine_QueueTest(engine, test_capture_screenshot);
}


int main(int, char**)
{
    HelloImGui::RunnerParams runnerParams;
    runnerParams.useImGuiTestEngine = true;

    runnerParams.callbacks.ShowGui = AppGui;
    runnerParams.callbacks.RegisterTests = MyRegisterTests;
    HelloImGui::Run(runnerParams);
}

#else // #ifdef IMGUI_BUNDLE_WITH_TEST_ENGINE
int main(int, char**) {}
#endif
